sos/utilities.py

```diff
@@ -226,22 +228,27 @@ def _check_poller(proc):
         else:
             reader = FakeReader(p, binary)
 
-        if poller:
-            while reader.running:
-                _check_poller(p)
-        else:
-            try:
+        try:
+            if poller:
+                while reader.running:
+                    _check_poller(p)
+            else:
                 # override timeout=0 to timeout=None, as Popen will treat the
                 # former as a literal 0-second timeout
                 p.wait(timeout if timeout else None)
-            except Exception:
-                p.terminate()
-                if to_file:
-                    _output.close()
-                # until we separate timeouts from the `timeout` command
-                # handle per-cmd timeouts via Plugin status checks
-                return {'status': 124, 'output': reader.get_contents(),
-                        'truncated': reader.is_full}
+        except ChildDeadlockError:
+            if to_file:
+                _output.close()
+            return {'status': 124, 'output': 'Error,Child process deadlock',
+                    'truncated': False}
```

Comment:
It is entirely possible for us to receive usable output up until the point a specific command deadlocks. This would eliminate that output, which would not be helpful to the person(s) reviewing an sos report.

In the case of a deadlocked child process, the parent process (the one we actually directly care about) still terminates via `timeout`, so it should all still be handled by the existing exception handler.

Reply:
In existing exception handler, The reader.get_contents() is invoked.  The 'sos report' will stuck in the `while self.running:`
https://github.com/sosreport/sos/blob/main/sos/utilities.py#L462

This is why I create the ChildDeadlockError, 
If command is timeout, reader.get_contents() can work.
if command have deadlock child process. reader.get_contents() will execute forever.

Reply:
`get_contents()` blocks on the internal `running` status, but that is entirely controlled by us. In the exception handling we can just directly set `reader.running = False`.

Reply:
I just modify my pull request as your ask, please check.